### ISSUE TITLE
Look into null values for PROTOCOL

### ISSUE THREAD

**Issue by rmarx** [rum-archive/rum-archive#24]
The PROTOCOL field in `rumarchive_page_loads` often has a value of `null`. 
This has improved somewhat over time, with it being about 33.5% in Oct 2021 and 20.5% in Dec 2023, but that's still very high.

I first thought this might be due to the `nextHopProtocol` field not being supported in all browsers, but according to this https://developer.mozilla.org/en-US/docs/Web/API/PerformanceResourceTiming/nextHopProtocol#browser_compatibility it should be broadly supported. 

That link also mentions needing a Timing-Allow-Origin, but even then it should give an empty string `""` value instead of actual null. Similarly, it's different from the "unknown" value (which has a quite low incidence overall). 

The query I'm using:

```
SELECT DATE as date,        PROTOCOL as protocol,         COUNT(*) as rowcount,         SUM(BEACONS) as beaconcount        FROM `akamai-mpulse-rumarchive.rumarchive.rumarchive_page_loads`         WHERE DATE = "2021-10-01" OR DATE = "2021-11-01" OR DATE = "2021-12-01" OR DATE = "2022-01-01" OR DATE = "2022-02-01" OR DATE = "2022-03-01" OR DATE = "2022-04-01" OR DATE = "2022-05-01" OR DATE = "2022-06-01" OR DATE = "2022-07-01" OR DATE = "2022-08-01" OR DATE = "2022-09-01" OR DATE = "2022-10-01" OR DATE = "2022-11-01" OR DATE = "2022-12-01" OR DATE = "2023-01-01" OR DATE = "2023-02-01" OR DATE = "2023-03-01" OR DATE = "2023-04-01" OR DATE = "2023-05-01" OR DATE = "2023-06-01" OR DATE = "2023-07-01" OR DATE = "2023-08-01" OR DATE = "2023-09-01" OR DATE = "2023-10-01" OR DATE = "2023-11-01" OR DATE = "2023-12-01"        GROUP BY DATE, PROTOCOL ORDER BY DATE ASC, PROTOCOL ASC
```

**Comment by nicjansma** [comment]
Yeah and according to [caniuse.com](https://caniuse.com/mdn-api_performanceresourcetiming_nexthopprotocol) global market share that supports `nextHopProtocol` should be more than 96%.  We're seeing NULL at ~20%, so almost 5x expected.

`Timing-Allow-Origin` restrictions shouldn't apply here, since we're only looking at the protocol of the base page (which is always by definition "same origin").

In addition, I would expect `nextHopProtocol` to be even more than 96% since if we don't collect it from the browser, we're able to collect it from the Akamai CDN for mPulse in many cases.

As far as collecting this data goes, we added `nextHopProtocol` support in boomerang 1.495+, which came out in July 2017.    There should be very little data coming in from boomerang.js versions that don't support it.

Let's look at the breakdown of protocols for a single day (same as the query above):

```sql
-- httpProtocol in data
SELECT  PROTOCOL as protocol,
        COUNT(*) as rowcount,
        SUM(BEACONS) as beaconcount,
        ROUND(SUM(BEACONS) /
          (SELECT SUM(BEACONS) FROM `akamai-mpulse-rumarchive.rumarchive.rumarchive_page_loads` WHERE DATE = '2023-12-01')
          * 100, 1) as beaconcountpct,
FROM    `akamai-mpulse-rumarchive.rumarchive.rumarchive_page_loads`
WHERE   DATE = "2023-12-01"
GROUP BY PROTOCOL
ORDER BY PROTOCOL ASC
;
```

<img width="660" alt="image" src="https://github.com/rum-archive/rum-archive/assets/1004649/3ea8a43c-b7db-4db9-b06c-56db7104bf71">

This matched Robin's query above showing ~20% NULL.

Next, let's look at the distribution of browser families that have NULL protocol data.  Maybe one particular browser is causing issues?

```sql
SELECT  UserAgentFamily,
        COUNT(*) as RowCount,
        SUM(BEACONS) as BeaconCount,
        ROUND(SUM(BEACONS) /
          (SELECT SUM(BEACONS)
            FROM `akamai-mpulse-rumarchive.rumarchive.rumarchive_page_loads`
            WHERE DATE = '2023-12-01' AND PROTOCOL IS NULL)
          * 100, 1) as BeaconCountPct,
FROM    `akamai-mpulse-rumarchive.rumarchive.rumarchive_page_loads`
WHERE   DATE = "2023-12-01"
    AND PROTOCOL IS NULL
GROUP BY UserAgentFamily
ORDER BY BeaconCountPct DESC
;
```

<img width="639" alt="image" src="https://github.com/rum-archive/rum-archive/assets/1004649/1e5efbaf-9a08-47d8-9c18-7743a173c3e2">

vs. the overall distribution:

<img width="636" alt="image" src="https://github.com/rum-archive/rum-archive/assets/1004649/73a0e5f5-9e86-456e-8cca-3c99c8da88fa">

Mobile Safari seems overly skewed in not having Protocol data (25.2% of overall data but 32.5% of NULL data).

I've looked at a bunch of raw data from Mobile Safari where the `nextHopProtocol` for the navigation was null, and I'm not seeing anything obvious yet.

Continuing to look into this!

**Comment by nicjansma** [comment]
I took a look at a few raw mPulse beacons e.g. focusing on Mobile Safari 17 on OSX 10.15, and comparing beacons that contain `nextHopProtocol` vs. those that don't.

Something I'm seeing on the beacons with NULL `nextHopProtocol` is that we also could **not** collect `transferSize`, `encodedBodySize` or `decodedBodySize` either, which should be supported in all Safari 17.

Those additional fields are **also** protected by same-origin restrictions.

So one guess is that Safaris' same-origin restrictions for these fields is being overly aggressive in some cases?

**Comment by bluesmoon** [comment]
I'm seeing bots pretending to be Mobile Safari. Might have more data in a
week or so.

**Comment by rmarx** [comment]
Thanks both; hopefully further research will turn up the real reason soon :) 

For others that might read this: the `protocol` field in the `rumarchive_resources` table is always `null` for now. This is a known current limitation however of how the data is piped throughout mPulse and this will hopefully change in the near future.

**Comment by nicjansma** [comment]
TODO: Look into percentages of the following scenarios that may have a null `nextHopProtocol`:

* Boomerang versions less than 1.495
* BFCache navs
* Navigations to HTML-cached pages

**Comment by nicjansma** [comment]
For the questions above, I looked at stats for the Top 100 mPulse apps included in the 2024-06-30 data:

* % of beacons from Boomerang versions less than 1.495: 0% (=0, all Top 100 apps are on at least 1.500)
* BFCache navigations: 0.4%
* Navigations to HTML-cached pages: 10.1% (in 29% of the page loads we can't tell either-way, as the `transferSize` attribute is missing)

I think the last bucket there (navigation to HTML-cached pages) is causing the biggest discrepancy in the data, especially w.r.t. Safari, where we see NHP empty a lot of the time.

**Repro**: this test page reports `nextHopProtocol` and `transferSize` for the NavigationTiming entry: https://dev.nicj.net/nexthopprotocol-transfersize.html

If you start out on the page, it should report `.nextHopProtocol=h2` and a `.transferSize` of 550 bytes or so.

![image](https://github.com/rum-archive/rum-archive/assets/1004649/b09e1bb5-a959-46ca-8706-f81c838a7704)

If you then click the `on domain 1` link, it navigates "back" to the exact same page, which should be in the HTTP cache.

Chrome then reports a `.transferSize=0` while `.nextHopProtocol=h2` (even though it was loaded from disk).  It reports this both for NavigationTiming and in the Developer Tools, where we see it's from the `(disk cache)` but `h2` still reported:

![image](https://github.com/rum-archive/rum-archive/assets/1004649/2c285183-31bf-4f6b-8e1c-082fae08113b)

Firefox shares the above behavior.

Safari however reports for the first navigation the same data: `.nextHopProtocol=h2` and a `.transferSize` of 550 bytes or so.

But then on the next navigation to the HTML-cached page, it shows `.transferSize=0` and `.nextHopProtocol=''`.

So there's some discrepancy in how the mainline browsers report `.nextHopProtocol` on a HTTP-cached page.

We've actually had a resource-timing issue about this in the past: https://github.com/w3c/resource-timing/issues/73

And a fix was merged in to an earlier version of the ResourceTiming spec to enforce that `nextHopProtocol` be empty when from cache: https://github.com/w3c/resource-timing/pull/78

However, with the refactor of ResourceTiming to align with Fetch, I'm not able to follow if that' still the spec'd behavior.  It would seem to me like it should be empty, but Chrome and Firefox currently reports `nextHopProtocol` of the original fetch.

I'll follow-up with the W3C WebPerf group to see what behavior we expect.

**Comment by nicjansma** [comment]
I've confirmed the expected behavior is `nextHopProtocol` should be empty for cached-based navigations.

So Safari is doing the correct behavior here, and I would expect the number of occurrences of `PROTOCOL` being empty to _increase_ if/once Chrome and Firefox fix this.

Filed:

* Chromium: https://issues.chromium.org/issues/350526044
* Firefox https://bugzilla.mozilla.org/show_bug.cgi?id=1905759
